nodes/stretch_ik_control.py

```diff
@@ -0,0 +1,1217 @@
+from pinocchio_ik_solver import PinocchioIKSolver
+from rclpy.action import ActionClient
+from rclpy.callback_groups import MutuallyExclusiveCallbackGroup
+from rclpy.duration import Duration
+from rclpy.node import Node
+from rclpy.task import Future
+from sensor_msgs.msg import JointState
+from std_srvs.srv import Trigger
+from stretch_body.robot_params import RobotParams
+from tf2_geometry_msgs import PoseStamped
+from tf_transformations import (
+    euler_from_quaternion,
+    quaternion_inverse,
+    quaternion_multiply,
+)
+from trajectory_msgs.msg import JointTrajectoryPoint
+
+
+class TerminationCriteria(Enum):
+    """
+    The termination criteria for the inverse Jacobian controller.
+    """
+
+    # Terminate when the commanded velocity is (near) zero.
+    ZERO_VEL = "zero_vel"
+    # Terminate when the error between the end-effector pose and the goal pose is (near) zero.
+    # NOTE: This will not converge if the goal pose is unreachable.
+    ZERO_ERR = "zero_err"
+
+
+class MotionGeneratorRetval(Enum):
+    """
+    The return values for the motion generators.
+    """
+
+    CONTINUE = 0
+    SUCCESS = 1
+    FAILURE = 2
+
+
+class StretchIKControl:
+    """
+    The StretchIKControl class allows users to compute Stretch's FK and IK, command the robot using an inverse
+    Jacobian velocity controller, and command the arm to a desired joint configuration.
+    """
+
+    def __init__(
+        self,
+        node: Node,
+        tf_buffer: tf2_ros.Buffer,
+        urdf_path: str,
+        speed_profile: SpeedProfile = SpeedProfile.MAX,  # TODO: consider allowing users to set this in the message!
+    ):
+        """
+        Create the StretchIKControl object.
+
+        Parameters:
+        ----------
+        node: The ROS2 node that will run the controller.
+        tf_buffer: The tf2 buffer to use to transform poses.
+        urdf_path: The path to the URDF file for the robot.
+        speed_profile: The speed profile to use to get max velocities and accelerations.
+        """
+        # Store the parameters
+        self.node = node
+        self.tf_buffer = tf_buffer
+        self.urdf_path = urdf_path
+        self.speed_profile = speed_profile
+        self.initialized = False
+
+    def initialize(self):
+        """
+        Initialize the StretchIKControl object.
+        """
+        # Load the IK solver.
+        # We restrict IK to the base rotation, arm lift, and arm joints,
+        # to avoid non-ideal IK solutions that either take advantage of
+        # joint redundancies (e.g., base rotation and wrist yaw) or get
+        # stuck in the gimbal lock that occurs when the wrist pitch is
+        # -pi/2. Note that this also means we can only control the below
+        # joints with inverse jacobian control.
+        self.controllable_joints = [
+            Joint.BASE_ROTATION,
+            Joint.ARM_LIFT,
+            Joint.ARM_L0,  # The URDF uses the most distal of the telescoping joints
```

Comment:
cannot do wrist?

Reply:
Not via inverse jacobian velocity control. `controllable_joints` are the joints that can be moved by inverse jacobian velocity control. I'll clarify that in a comment.

In practice, because the wrist also impacts the same end effector pose dimensions as the above three, we were getting some Jacobians that were attempting to use them to compensate for each other, e.g., trying to pitch up the wrist if arm lift overshot.